src/Backend/Haskell.idr

```diff
@@ -710,32 +805,39 @@ decodeDef {n} t@(TName tname td) =
          pure (HsInt (finToInteger i), HsDo $ (map (\ (x, e) => (Just x, e)) args')++[(Nothing, hsReturn (HsInn name (map fst args')))])
 
     -- Idris is not clever enough to figure out the following type if written as a case expression
-    genCase : (n : Nat) -> HsType -> HsTerm -> Vect n (HsType, HsTerm) -> TDef n -> HsTerm
+    genCase : (n : Nat) -> HsType -> HsTerm -> Vect n (HsType, HsTerm) -> TDef n -> HaskellLookupM HsTerm
     genCase n currType currTerm env (TMu [(name, td)]) =
-      simplify $ snd $ runTermGen ((currType, currTerm)::env) (genCases {k = S Z} (S n) FZ (name, td))
-    genCase n currType currTerm env (TMu {k} tds)      = runTermGen ((currType, currTerm)::env) $
-      do cases <- mapWithIndexA (genCases (S n)) tds
-         [i] <- freshVars 1 "i"
-         pure $ HsDo [ (Just i, hsReadInt (fromNat k))
-                     , (Nothing, simplify $ hsCaseDef i (toList cases) hsFailDecode)
-                     ]
-    genCase n currType currTerm env td = simplify $ runTermGen env (decode td)
+      toHaskellLookupM $ map (simplify . snd) $ runTermGen ((currType, currTerm)::env) (genCases {k = S Z} (S n) FZ (name, td))
+    genCase n currType currTerm env (TMu {k} tds)      =
+      toHaskellLookupM $ runTermGen ((currType, currTerm)::env) (
+        do cases <- traverseWithIndex (genCases (S n)) tds
+           [i] <- freshVars 1 "i"
+           pure $ HsDo [ (Just i, hsReadInt (fromNat k))
+                       , (Nothing, simplify $ hsCaseDef i (toList cases) hsFailDecode)
+                       ])
+    genCase n currType currTerm env td = toHaskellLookupM $ map simplify $ runTermGen env (decode td)
 
 ASTGen Haskell HsType True where
-  msgType  (Unbounded tn) = makeType' freshEnv tn
-  generateTyDefs tns =
-    evalState
-      (foldlM (\lh,(Unbounded tn) => (lh ++) <$> (makeDefs' tn)) [] tns)
-      (the (List Name) [])
-  generateTermDefs tns =
-    evalState
-      (foldlM (\lh, (Unbounded {n} tn) => (lh ++) <$>
-                  let genFrom = dependencies freshEnv (def tn) ++ [(n ** tn)] in
-                  foldlM (\lh', (_ ** tm) =>
-                            (lh' ++) <$> ifNotPresent (name tm) (pure [encodeDef tm, decodeDef tm]))
-                         [] genFrom)
-              [] tns)
-      (the (List Name) [])
+  msgType  (Unbounded tn) = runHaskellLookupM $ makeType' freshEnv tn
+  generateTyDefs tns = runMakeDefsM $ concat <$> traverseEffect (\(Unbounded tn) => makeDefs' tn) (toVect tns)
+  generateTermDefs tns = runMakeDefsM $ 
+    do gen <- traverseEffect genHaskell (toVect tns)
+       pure $ concat gen
+    where
+      genTerms : TNamed n -> HaskellLookupM (List Haskell)
+      genTerms tn = pure [!(encodeDef tn), !(decodeDef tn)]
+
+      generateNext : TNamed n -> HaskellDefM (List Haskell)
+      generateNext tn = if (name tn) `elem` !('Names :- get)
```

Comment:
Isn't this repeating the definition of `ifNotPresent`?

Reply:
Indeed, fixed